### PR TITLE
allow parameters on actions to be treated as final

### PR DESCRIPTION
**Pull request by rabbah**
This pull request allows actions to declare their parameters "final" and may not be overridden by invoke-time parameters.

### CODE REVIEW

File: common/scala/src/main/scala/whisk/core/entity/Parameter.scala [hunk 1]
```diff
@@ -53,26 +53,30 @@ protected[core] class Parameters protected[entity] (
             name.size + value.size
     }.foldLeft(0 B)(_ + _)
 
-    protected[entity] def ++(p: (ParameterName, ParameterValue)) = {
+    protected[entity] def +(p: (ParameterName, ParameterValue)) = {
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/Parameter.scala]: FWIW regarding scala patterns: The `apply` method for `Parameters` returns an `Option` vs. with `Map` `apply` is the dangerous version which will throw an exception. Having a `get` method instead could be desirable? (I stumbled over this at least once)

Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Parameter.scala]: Fair. I'll nuke it as I've removed almost all of them.

Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/Parameter.scala]: https://github.com/openwhisk/openwhisk/pull/1710/commits/12255b8a9fba9250227835f5b0c372c59e234c37.

File: common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala [hunk 2]
```diff
@@ -109,6 +109,14 @@ case class WhiskAction(
     require(exec != null, "exec undefined")
     require(limits != null, "limits undefined")
 
+    /** @return true iff action has appropriate annotation. */
+    def hasFinalParamsAnnotation = {
```
Comment (markusthoemmes) [inline common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala]: Want to fold `asBool` https://github.com/openwhisk/openwhisk/pull/1715/files#diff-85adc8c36c3cfcd17be782427443d12eR87 into this pull-request?

File: tests/src/whisk/core/controller/test/MetaApiTests.scala [hunk 3]
```diff
@@ -205,15 +212,20 @@ class MetaApiTests extends ControllerTestCommon with WhiskMetaApi with BeforeAnd
         implicit val tid = transid()
         val methods = Seq(Get, Post, Delete)
 
-        methods.map { m =>
+        methods.foreach { m =>
```
Comment (markusthoemmes) [inline tests/src/whisk/core/controller/test/MetaApiTests.scala]: ❤️

File: core/controller/src/main/scala/whisk/core/controller/Meta.scala [hunk 4]
```diff
@@ -105,15 +143,23 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
...
-                    "__ow_meta_verb" -> method.value.toLowerCase.toJson,
-                    "__ow_meta_path" -> restofPath.toJson,
-                    "__ow_meta_namespace" -> user.namespace.toJson)
-                invokeAction(_, action, Some(JsObject(content)), blocking = true, waitOverride = true)
+            systemKey flatMap { identity =>
+                val invokeParams = if (action.hasFinalParamsAnnotation) {
+                    requestParams -- action.parameters.immutableParameters
+                } else requestParams // in the absence of immutable annotations, return the request untouched
+
+                if (invokeParams.size == requestParams.size) {
```
Comment (markusthoemmes) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: If I understand this correctly, `invokeParams` is redundant. Can this be rewritten as:

```
// only allow execution iff request does not contain parameters declared as final
if(requestParams.keySet.intersect(action.parameters.immutableParameters).nonEmpty) ...
```

I think its more self-explanatory than comparing the sizes? What do you think?

**Comment by markusthoemmes** [comment]
Don't we want this on **any** action? Currently it's only enforced in the Meta API?

File: common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala (hunk 2 above)
Comment (rabbah) [inline common/scala/src/main/scala/whisk/core/entity/WhiskAction.scala]: No. 😀

**Comment by rabbah** [comment]
We could make apply this to all parameters but it's arguable whether we want to consider all defined params immutable (vs invoke time prohibition only as in this or) so as not to allow overriding say between binding and package or action and package/binding. For that it's "experimental".  Adopting immutable (and losing default params with overrides in the process) consistently is appealing but also not backward comprible. Invoke time only (as in this pr) can be generalized to all actions though, as you noted.

File: core/controller/src/main/scala/whisk/core/controller/Meta.scala (hunk 4 above)
Comment (rabbah) [inline core/controller/src/main/scala/whisk/core/controller/Meta.scala]: Yes but intersect could be more expensive so I opted for a size check instead against the original map.

**Comment by rabbah** [comment]
sorry about that - pgs 3/139 and 3/143.

**Comment by psuter** [comment]
Merged as 0eaee0463103e416d12ef07fbe736a9f344ea5e9

**Pull request closed without merging**